### ISSUE TITLE
Overly large stack size in pj_init:get_opt

### ISSUE THREAD

**Issue by schwehr** [OSGeo/PROJ#697]
For any heavily threaded program, large stack frames are a real problem.  get_opt allocates 64K on the stack.  Any way that `sword` could be moved to the heap?

```
/* Maximum size of files using the "escape carriage return" feature */
#define MAX_CR_ESCAPE 65537
typedef struct {
    projCtx ctx;
    PAFile fid;
    char buffer[MAX_CR_ESCAPE];
    int buffer_filled;
    int at_eof;
} pj_read_state;

// ...
static paralist *
get_opt(projCtx ctx, paralist **start, PAFile fid, char *name, paralist *next,
        int *found_def) {
    pj_read_state *state = (pj_read_state*) calloc(1,sizeof(pj_read_state));
    char sword[MAX_CR_ESCAPE];  // <---- Ouch
```

See also: (C++ based, so only somewhat relevant)

- http://goo.gl/vuA3D6 - Use vector<T>(length, initial_value) for local blocks of storage
- https://trac.osgeo.org/gdal/ticket/5748

**Comment by busstoptaktik** [comment]
@schwehr: PR #693 adresses this issue, and provides numerous other improvements to a corner of the code base that was rather tough to maintain and improve.

I would appreciate a review of the PR from you: It is rather intrusive, but also brings big benefits, so having a few extra eyes on the code would be beneficial.

**Comment by schwehr** [comment]
reviewing now

**Comment by busstoptaktik** [comment]
Thanks. Note that I have just now pushed a minor (1 line) correction in pj_internal.c, in an attempt to eliminate an adressing underflow error.

**Comment by schwehr** [comment]
@busstoptaktik Hope my comments helped at least some.  I didn't see any changes that impacted `MAX_CR_ESCAPE`.  It wasn't clear that your code address this issue.  I presume there is some follow up change that will address the large on the stack buffer issue.  Yes?

**Comment by busstoptaktik** [comment]
@schwehr Thanks for some very useful comments, that it will take me some time to address all of.

Some of the comments (most notably avoiding the use of 0 for pointers, limiting linelength to that of classic 80x24 VDUs, and considering the use of non-void destructors) are matters of general style, and I will have to consider them more at depth in a more general context than this PR - but your efforts are nevertheless much appreciated.

Regarding not seeing changes that impacted `MAX_CR_ESCAPE`, you probably didn't see them because there is nothing left to impact: `MAX_CR_ESCAPE` is gone, and with that, so are large stack allocations

The largest stack allocation after this PR is around 1100 bytes (at least in the corner of the code touched here).

The `get_opt` function, where the large stack allocation took place is entirely gone.

**Comment by busstoptaktik** [comment]
@schwehr I think I have now addressed all your comments to PR #693. So a huge thank you for your time and efforts in reviewing - it has clearly led to much improvement of the code.

Everything builds and tests flawlessly now, and I intend to merge the PR in a few hours. When merged, I will also close this issue. If you do not agree that it has been sufficiently addressed by PR #693, please reopen with a comment saying so.

**Comment by schwehr** [comment]
I'm stuck back a ways in the patches, so I will certainly reopen if I run into it again.  Thanks for the hard work!
